lua/lib/lfospec.lua

```diff
@@ -0,0 +1,449 @@
+-- @tparam[opt] string mode How to advance the LFO (options: 'clocked', 'free'; default: 'clocked')
+-- @tparam[opt] number period The timing of this LFO's advancement. If mode is 'clocked', argument is in beats. If mode is 'free', argument is in Hz.
+-- @tparam[opt] function fn A callback function to perform as the LFO advances. This library passes both the scaled and the raw value to the callback function.
+function LFO.new(shape, min, max, depth, mode, period, fn)
+  local i = {}
+  setmetatable(i, LFO)
+  i.scaled = 0
+  i.raw = 0
+  i.phase_counter = 0
+  i.shape = shape == nil and 'sine' or shape
+  i.min = min == nil and 0 or min
+  i.max = max == nil and 1 or max
+  i.depth = depth == nil and 1 or depth
+  i.enabled = 0
+  i.mode = mode == nil and 'clocked' or mode
+  i.period = period == nil and 4 or period
+  i.reset_target = 'floor'
+  i.baseline = 'min'
+  i.offset = 0
+  i.ppqn = 96
+  i.controlspec = {
+    warp = 'linear',
+    step = 0.01,
+    units = '',
+    quantum = 0.01,
+    wrap = false,
+    formatter = nil
+  }
+  i.action = fn == nil and (function(scaled, raw) end) or fn
+  return i
+end
+
+-- PARAMETERS UI /
+local function lfo_params_visibility(state, i)
+  params[state](params, "lfo_baseline_"..i)
+  params[state](params, "lfo_offset_"..i)
+  params[state](params, "lfo_depth_"..i)
+  params[state](params, "lfo_scaled_"..i)
+  params[state](params, "lfo_raw_"..i)
+  params[state](params, "lfo_mode_"..i)
+  if state == "show" then
+    if params:get("lfo_mode_"..i) == 1 then
+      params:hide("lfo_free_"..i)
+      params:show("lfo_clocked_"..i)
+    elseif params:get("lfo_mode_"..i) == 2 then
+      params:hide("lfo_clocked_"..i)
+      params:show("lfo_free_"..i)
+    end
+  else
+    params:hide("lfo_clocked_"..i)
+    params:hide("lfo_free_"..i)
+  end
+  params[state](params, "lfo_shape_"..i)
+  params[state](params, "lfo_min_"..i)
+  params[state](params, "lfo_max_"..i)
+  params[state](params, "lfo_reset_"..i)
+  params[state](params, "lfo_reset_target_"..i)
+  _menu.rebuild_params()
+end
+
+local function build_lfo_spec(lfo,i,bound)
+  params:add{
+    type = 'control',
+    id = 'lfo_'..bound..'_'..i,
+    name = 'lfo '..bound,
+    controlspec = controlspec.new(
+      lfo.min,
+      lfo.max,
+      lfo.controlspec.warp,
+      lfo.controlspec.step,
+      bound == 'min' and lfo.min or lfo.max,
+      lfo.controlspec.units,
+      lfo.controlspec.quantum,
+      lfo.controlspec.wrap
+    ),
+    formatter = lfo.controlspec.formatter
+  }
+  params:set_action('lfo_'..bound..'_'..i, function(x)
+    lfo:set(bound,x)
+  end)
+end
+
+local function lfo_bang(i)
+  local function lb(prm)
+    params:lookup_param("lfo_"..prm.."_"..i):bang()
+  end
+  lb('depth')
+  lb('min')
+  lb('max')
+  lb('baseline')
+  lb('offset')
+  lb('mode')
+  lb('clocked')
+  lb('free')
+  lb('shape')
+  lb('reset')
+  lb('reset_target')
+end
+
+-- / PARAMETERS UI
+
+-- SCRIPTING /
+
+local function process_lfo(id)
```

Comment:
generally this function needs to be combed through to make sure you're not doing math-work that doesn't need to be done since there is a lot of per-mode calculation

Reply:
+1. for example...

Reply:
oh duhhhh, thank you!